### PR TITLE
webdataset prototype - LoadFilesFromDiskIterableDataset

### PR DESCRIPTION
**Pull request by glaringlee**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #49353 webdataset prototype - GroupByFilenameIterableDataset
* #49140 webdataset prototype - RoutedDecoderIterableDataset
* #49026 webdataset prototype - ReadFilesFromZipIterableDataset
* #49000 webdataset prototype - ReadFilesFromTarIterableDataset
* **#48955 webdataset prototype - LoadFilesFromDiskIterableDataset**
* #48944 webdataset prototype - ListDirFilesIterableDataset

Differential Revision: [D25541393](https://our.internmc.facebook.com/intern/diff/D25541393)

### CODE REVIEW

File: test/test_dataset.py [hunk 1]
```diff
@@ -3,14 +3,15 @@
...
-    temp_dir = tempfile.TemporaryDirectory()
+    # Note: the temp dir and files within it will be deleted in tearDown()
+    temp_dir = tempfile.TemporaryDirectory()  # noqa: P201
     temp_dir_path = temp_dir.name
-    temp_file1 = tempfile.NamedTemporaryFile(dir=temp_dir_path, delete=False)
-    temp_file2 = tempfile.NamedTemporaryFile(dir=temp_dir_path, delete=False)
-    temp_file3 = tempfile.NamedTemporaryFile(dir=temp_dir_path, delete=False)
+    temp_file1 = tempfile.NamedTemporaryFile(dir=temp_dir_path, delete=False)  # noqa: P201
+    temp_file2 = tempfile.NamedTemporaryFile(dir=temp_dir_path, delete=False)  # noqa: P201
+    temp_file3 = tempfile.NamedTemporaryFile(dir=temp_dir_path, delete=False)  # noqa: P201
```
Comment (VitalyFedyunin) [inline test/test_dataset.py:14]: Can you please commend this noqa

File: test/test_dataset.py [hunk 2]
```diff
@@ -34,5 +35,14 @@ def test_listdirfiles_iterable_dataset(self):
         for pathname in dataset:
             self.assertTrue(pathname in self.temp_files)
 
+    def test_loadfilesfromdisk_iterable_dataset(self):
+        temp_dir = self.temp_dir.name
+        dataset1 = ListDirFilesIterableDataset(temp_dir, '')
+        dataset2 = LoadFilesFromDiskIterableDataset(dataset1)
+
+        for rec in dataset2:
```
Comment (VitalyFedyunin) [inline test/test_dataset.py:43]: Are you sure that traverse order is guaranteed?

File: torch/utils/data/datasets/common.py [hunk 3]
```diff
@@ -35,3 +35,17 @@ def onerror(err : OSError):
...
+    if not isinstance(pathnames, Iterable):
+        warnings.warn("get_file_binaries_from_pathnames needs the input be an Iterable")
+        raise TypeError
+
+    for pathname in pathnames:
+        if not isinstance(pathname, str):
+            warnings.warn("file pathname must be string type, but got {}".format(type(pathname)))
+            raise TypeError
+
+        yield (pathname, open(pathname, 'rb'))
```
Comment (VitalyFedyunin) [inline torch/utils/data/datasets/common.py:51]: Can you please verify that we are releasing file handler with created stream objects gets garbage collected

File: test/test_dataset.py (hunk 1 above)
Comment (glaringlee) [reply test/test_dataset.py:14]: These #noqa: P201 is removed in the later stack PRs, i changed another way to create these files.

File: test/test_dataset.py (hunk 2 above)
Comment (glaringlee) [reply test/test_dataset.py:43]: Traverse order is guaranteed in LoadFilesFromDiskIterableDataset.
I think the question is whether the traverse order is guaranteed in ListDirFilesIterableDataset.
ListDirFilesIterableDataset use os.walk, currently no enforced order, should we enforce some kind of order? sort by file path+name?

File: torch/utils/data/datasets/common.py (hunk 3 above)
Comment (glaringlee) [reply torch/utils/data/datasets/common.py:51]: Ah, good catch. I don't think all the platform will do this releasing automatically. I will update the consumer dataset (tardataset, decoder dataset, etc) to try explicitly release it if possible.

Comment (VitalyFedyunin) [reply torch/utils/data/datasets/common.py:51]: Alternatively wrap (or find wrapper) which will take care of closing stream on object __del__

File: test/test_dataset.py (hunk 2 above)
Comment (VitalyFedyunin) [reply test/test_dataset.py:43]: No need to order it.

File: torch/utils/data/datasets/common.py [hunk 4]
```diff
@@ -2,6 +2,41 @@
 import fnmatch
 import warnings
 from typing import List, Union, Iterable
+from io import BufferedIOBase
+
+class StreamWrapper:
```
Comment (VitalyFedyunin) [inline torch/utils/data/datasets/common.py:7]: Need to be Child of Stream class or similar, to make sure that isinstance work correctly.

Comment (glaringlee) [reply torch/utils/data/datasets/common.py:7]: Double checked the BufferedIOBase which is used as streaming handle. It inherited from IOBase which has __del__ implemented calling close() already. I'll remove this StreamWrapper, no longer needed.

File: torch/utils/data/datasets/common.py (hunk 3 above)
Comment (glaringlee) [reply torch/utils/data/datasets/common.py]: Double checked the impl of BufferedIOBased class (which is used as streaming handle obj), it inherited from IOBased which has __del__ implemented calling close() already, we don't need to take care of this manually.

**Review by VitalyFedyunin — APPROVED** [review]

**Pull request closed without merging**